src/combinatorics/simple_graph/inc_matrix.lean

```diff
@@ -0,0 +1,415 @@
+λ i e, if i = o.head e then (1 : R) else (if i = o.tail e then -1 else 0)
+
+@[simp]
+lemma oriented_inc_matrix_apply {i : V} {e : G.edge_set} :
+  G.oriented_inc_matrix R o i e = ite (i = o.head e) (1 : R) (ite (i = o.tail e) (-1) 0) := rfl
+
+lemma oriented_inc_matrix_head {i : V} {e : G.edge_set} (H_head : i = o.head e) :
+  G.oriented_inc_matrix R o i e = 1 :=
+by simp only [H_head, if_true, eq_self_iff_true, oriented_inc_matrix_apply]
+
+lemma oriented_inc_matrix_tail {i : V} {e : G.edge_set} (H_tail : i = o.tail e) :
+  G.oriented_inc_matrix R o i e = -1 :=
+by simp only [H_tail, oriented_inc_matrix, (G.head_neq_tail).symm,
+              if_false, if_true, eq_self_iff_true]
+
+lemma oriented_inc_matrix_zero {i : V} {e : G.edge_set} [char_zero R] :
+  G.oriented_inc_matrix R o i e = 0 ↔ i ≠ o.head e ∧ i ≠ o.tail e :=
+begin
+  by_cases H₁ : i = o.head e,
+  { simp only [oriented_inc_matrix, H₁, if_true, eq_self_iff_true, not_true,
+               ne.def, one_ne_zero, false_and] },
+  { by_cases H₂ : i = o.tail e,
+    { simp only [H₂, oriented_inc_matrix_tail, eq_self_iff_true, not_true,
+                 ne.def, neg_eq_zero, one_ne_zero, and_false] },
+    { simp only [H₁, H₂, eq_self_iff_true, if_false, ne.def,
+                 not_false_iff, and_self, oriented_inc_matrix_apply] } }
+end
+
+lemma oriented_inc_matrix_non_zero {i : V} {e : G.edge_set} [char_zero R] :
+  ¬ G.oriented_inc_matrix R o i e = 0 ↔ i = o.head e ∨ i = o.tail e :=
+begin
+  by_cases H₁ : i = o.head e,
+  { simp only [H₁, if_true, true_or, eq_self_iff_true, ne.def,
+               not_false_iff, one_ne_zero, oriented_inc_matrix_apply] },
+  { by_cases H₂ : i = o.tail e,
+    { simp only [H₂, oriented_inc_matrix_tail, eq_self_iff_true, ne.def, or_true,
+                 not_false_iff, neg_eq_zero, one_ne_zero] },
+    { simp only [H₁, H₂, eq_self_iff_true, not_true, if_false,
+                 ne.def, oriented_inc_matrix_apply, or_self] } }
+end
+
+-- 5. (N(o) i e) ^ 2 = M i e, for any orientation o, vertex i and edge e.
+theorem oriented_inc_matrix_elem_squared {i : V} {e : G.edge_set} [char_zero R] :
+  G.oriented_inc_matrix R o i e * G.oriented_inc_matrix R o i e = G.inc_matrix R i e :=
+begin
+  by_cases H_head : i = o.head e,
+  { rw [G.oriented_inc_matrix_head R H_head, H_head, mul_one, eq_comm, inc_matrix_one],
+    exact G.incidence_set_orientation_head },
+  { by_cases H_tail : i = o.tail e,
+    { rw [G.oriented_inc_matrix_tail R H_tail, H_tail, mul_neg_eq_neg_mul_symm, mul_one,
+          neg_neg, eq_comm, inc_matrix_one],
+      exact G.incidence_set_orientation_tail },
+    { rw [(G.oriented_inc_matrix_zero R).mpr ⟨H_head, H_tail⟩, mul_zero, eq_comm, inc_matrix_zero],
+      exact G.not_inc_set_orientation H_head H_tail } }
+end
+
+-- 6. For any adj. vertices i j and edge e, N(o) i e * N(o) j e = if e = ⟦(i,j)⟧ then -1 else 0.
+theorem oriented_inc_matrix_mul_of_adj {i j : V} {e : G.edge_set} (H_adj : G.adj i j) [char_zero R]:
```

Comment:
this proof looks horrific, is there really no better way?

Reply:
I am looking into it, I assume you don't like that I split the problem in so many cases, and many of them use approximately the same things.

Reply:
precisely, yeah

Reply:
I managed to reduce most of the repeated work, but I had to add new helper lemmas for that. Still think this approach is clearly better here :smile: